// rlog-js/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _Rlog_instances, _Rlog_genApi;
Object.defineProperty(exports, "__esModule", { value: true });
const chalk_1 = __importDefault(require("chalk"));
const fs = __importStar(require("fs-extra"));
const moment_1 = __importDefault(require("moment"));
require("moment-timezone");
/**
 * Configuration class for Rlog
 * Manages all logging behavior settings
 */
class Config {
    constructor() {
        this.enableColorfulOutput = true;
        this.logFilePath = undefined;
        this.timeFormat = "YYYY-MM-DD HH:mm:ss.SSS";
        this.timezone = undefined;
        this.joinChar = " ";
        this.blockedWordsList = [];
        this.screenLength = process.stdout.columns || 80;
        this.autoInit = true;
        this.silent = false;
        this.customColorRules = [
            {
                reg: "false",
                color: "red",
            },
            {
                reg: "true",
                color: "green",
            },
            {
                reg: "((2(5[0-5]|[0-4]\\d))|[0-1]?\\d{1,2})(\\.((2(5[0-5]|[0-4]\\d))|[0-1]?\\d{1,2})){3}",
                color: "cyan",
            },
            {
                reg: "[a-zA-z]+://[^\\s]*",
                color: "cyan",
            },
            {
                reg: "\\d{4}-\\d{1,2}-\\d{1,2}",
                color: "green",
            },
            {
                reg: "\\w+([-+.]\\w+)*@\\w+([-.]\\w+)*\\.\\w+([-.]\\w+)*",
                color: "cyan",
            },
            {
                reg: "[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}",
                color: "cyan",
            },
            {
                reg: "(w+)s*:s*([^;]+)",
                color: "cyan",
            },
        ];
        if (!this.enableColorfulOutput) {
            chalk_1.default.level = 0;
        }
    }
    /**
     * Update configuration for this instance only
     * @param obj - Partial configuration object
     * @example
     * ```typescript
     * rlog.config.setConfig({
     *   timezone: 'Asia/Shanghai',
     *   silent: true
     * });
     * ```
     */
    setConfig(obj) {
        if (!obj)
            return;
        for (const key in obj) {
            if (obj.hasOwnProperty(key)) {
                this[key] = obj[key];
            }
        }
    }
    /**
     * Update configuration globally for all instances
     * Changes the default prototype values
     * @param obj - Partial configuration object
     * @example
     * ```typescript
     * rlog.config.setConfigGlobal({
     *   blockedWordsList: ['password', 'secret']
     * });
     * ```
     */
    setConfigGlobal(obj) {
        if (!obj)
            return;
        for (const key in obj) {
            if (obj.hasOwnProperty(key)) {
                const value = obj[key];
                this[key] = value;
                Config.prototype[key] = value;
            }
        }
    }
}
/**
 * Toolkit class providing utility functions for log formatting
 */
class Toolkit {
    constructor(config) {
        this.config = config;
    }
    /**
     * Ensure log file exists, create if necessary
     * @param path - Path to log file
     */
    async checkLogFile(path) {
        try {
            fs.ensureFileSync(path);
            await fs.promises.access(path, fs.constants.F_OK);
        }
        catch (err) {
            try {
                await fs.promises.writeFile(path, "");
            }
            catch (err) {
                this.screen.error("Could not create file, error: " + err);
            }
        }
    }
    /**
     * Apply custom color rules to colorize matched patterns in string
     * Preserves existing ANSI color codes
     * @param str - String to colorize
     * @returns Colorized string with ANSI codes
     */
    colorizeString(str) {
        if (!str || typeof str !== "string")
            return str;
        const ansiColorRegex = /(\u001b\[\d+m)/g;
        const parts = str.split(ansiColorRegex);
        let activeColorStack = [];
        const result = [];
        for (const part of parts) {
            if (part.startsWith("\u001b[")) {
                if (part === "\u001b[39m") {
                    activeColorStack = [];
                }
                else {
                    activeColorStack.push(part);
                }
                result.push(part);
                continue;
            }
            if (!part)
                continue;
            const currentColorState = [...activeColorStack];
            let processedText = part;
            for (const { reg, color } of this.config.customColorRules) {
                const regex = new RegExp(reg, "g");
                processedText = processedText.replace(regex, (match) => {
                    const chalkFn = chalk_1.default[color];
                    if (typeof chalkFn !== 'function') {
                        return match;
                    }
                    const coloredMatch = chalkFn(match);
                    const colorParts = coloredMatch.split(ansiColorRegex).filter(Boolean);
                    const colorStart = colorParts[0];
                    const matchText = colorParts
                        .filter((p) => !p.startsWith("\u001b["))
                        .join("");
                    const restoreColor = currentColorState.length > 0
                        ? currentColorState.join("")
                        : "\u001b[39m";
                    return colorStart + matchText + restoreColor;
                });
            }
            result.push(processedText);
        }
        return result.join("");
    }
    /**
     * Format current time according to config
     * @returns Formatted time string or timestamp number
     * @example
     * ```typescript
     * formatTime() // "2025-12-18 10:30:00.123"
     * // With timeFormat: 'timestamp'
     * formatTime() // 1734501000123
     * ```
     */
    formatTime() {
        const now = (0, moment_1.default)();
        const { timeFormat, timezone } = this.config;
        if (timeFormat === "timestamp") {
            return now.valueOf();
        }
        if (timezone) {
            switch (timeFormat) {
                case "ISO":
                    return now.tz(timezone).toISOString();
                case "GMT":
                    return now.tz("GMT").format("YYYY-MM-DDTHH:mm:ss.SSS") + "Z";
                case "UTC":
                    return now.utc().format();
                default:
                    return now.tz(timezone).format(timeFormat);
            }
        }
        return now.format(timeFormat);
    }
    /**
     * Mask sensitive content in string using blocked words list
     * Supports both literal strings and regex patterns
     * @param str - String to process
     * @returns String with sensitive content replaced by asterisks
     * @example
     * ```typescript
     * // With blockedWordsList: ['password', '[0-9]{9}']
     * encryptPrivacyContent('password: 123456789')
     * // Returns: "********: *********"
     * ```
     */
    encryptPrivacyContent(str) {
        if (typeof str !== "string" || !this.config.blockedWordsList?.length) {
            return str;
        }
        if (!this._regexCache) {
            this._regexCache = this.config.blockedWordsList.map((pattern) => {
                try {
                    return new RegExp(pattern, "g");
                }
                catch (e) {
                    return new RegExp(pattern.replace(/[-\/\\^$*+?.()|[\]{}]/g, "\\$&"), "g");
                }
            });
        }
        return this._regexCache.reduce((result, regex) => {
            return result.replace(regex, (match) => {
                return "*".repeat(match.length);
            });
        }, str);
    }
    /**
     * Colorize variable based on its type
     * @param variable - Variable to colorize
     * @returns Colored string representation
     */
    colorizeType(variable) {
        if (variable === null)
            return chalk_1.default.red("null");
        if (variable === undefined)
            return chalk_1.default.gray("undefined");
        const type = typeof variable;
        switch (type) {
            case "string":
                return variable;
            case "number":
                return chalk_1.default.blue(variable.toString());
            case "boolean":
                return variable ? chalk_1.default.green("true") : chalk_1.default.red("false");
            case "object":
                try {
                    if (Array.isArray(variable)) {
                        return chalk_1.default.yellow(JSON.stringify(variable, null, 2));
                    }
                    return chalk_1.default.magenta(JSON.stringify(variable, null, 2));
                }
                catch (e) {
                    return chalk_1.default.red("[Circular Object]");
                }
            case "function":
                return chalk_1.default.cyan(variable.toString().split("\n")[0] + "...");
            case "symbol":
                return chalk_1.default.yellow(variable.toString());
            default:
                return String(variable);
        }
    }
    /**
     * Add padding to multi-line strings (indent lines after first)
     * @param str - String to pad
     * @param width - Number of spaces to add
     * @returns Padded string
     */
    padLines(str, width) {
        if (!str)
            return "";
        str = String(str);
        if (!str.includes("\n"))
            return str;
        const lines = str.split("\n");
        const padding = " ".repeat(width);
        return (lines[0] +
            "\n" +
            lines
                .slice(1)
                .map((line) => padding + line)
                .join("\n"));
    }
    /**
     * Convert any object to string representation
     * @param obj - Object to stringify
     * @returns String representation
     */
    stringify(obj) {
        if (typeof obj === "string") {
            return obj;
        }
        if (typeof obj === "object") {
            return JSON.stringify(obj, null, 2);
        }
        return obj.toString();
    }
}
/**
 * Screen class for console output
 * Handles all terminal/stdout logging with colors and formatting
 */
class Screen {
    constructor(toolkit) {
        this.toolkit = toolkit;
    }
    /**
     * Format a log message with timestamp and type
     * @private
     */
    _formatMessage(type, color, message, time) {
        const timeheader = `[${time || this.toolkit.formatTime()}]`;
        const chalkFn = chalk_1.default[color];
        const colorizedType = typeof chalkFn === 'function' ? chalkFn(type) : type;
        const processedMessage = this.toolkit.encryptPrivacyContent(this.toolkit.padLines(type === "SUCC" || type === "EXIT"
            ? (typeof chalkFn === 'function' ? chalkFn(message) : message)
            : this.toolkit.colorizeType(message), timeheader.length + 7));
        return `${timeheader}[${colorizedType}] ${this.toolkit.colorizeString(processedMessage)}\n`;
    }
    /**
     * Write log to stdout
     * @private
     */
    _log(type, color, message, time) {
        process.stdout.write(this._formatMessage(type, color, message, time));
    }
    /**
     * Log an information message to console
     * @param message - Message to log
     * @param time - Optional custom timestamp
     */
    info(message, time) {
        this._log("INFO", "cyan", message, time);
    }
    /**
     * Log a warning message to console
     * @param message - Message to log
     * @param time - Optional custom timestamp
     */
    warning(message, time) {
        this._log("WARN", "yellow", message, time);
    }
    /**
     * Log a warning message to console (alias for warning)
     * @param message - Message to log
     * @param time - Optional custom timestamp
     */
    warn(message, time) {
        this._log("WARN", "yellow", message, time);
    }
    /**
     * Log an error message to console
     * @param message - Message to log
     * @param time - Optional custom timestamp
     */
    error(message, time) {
        this._log("ERR!", "red", message, time);
    }
    /**
     * Log a success message to console
     * @param message - Message to log
     * @param time - Optional custom timestamp
     */
    success(message, time) {
        this._log("SUCC", "green", message, time);
    }
    /**
     * Log an exit message to console
     * @param message - Message to log
     * @param time - Optional custom timestamp
     */
    exit(message, time) {
        this._log("EXIT", "red", message, time);
    }
}
/**
 * File class for writing logs to file system
 * Manages log file stream and async writes
 */
class File {
    constructor(toolkit, config, screen) {
        /** Write stream for log file */
        this.logStream = null;
        this.toolkit = toolkit;
        this.config = config;
        this.screen = screen;
        if (this.config.autoInit)
            this.init();
    }
    /**
     * Initialize log file and create write stream
     * Called automatically if autoInit is true
     */
    init() {
        if (this.config.logFilePath && !this.logStream) {
            this.toolkit.checkLogFile(this.config.logFilePath);
            try {
                this.logStream = fs.createWriteStream(this.config.logFilePath, {
                    flags: "a",
                });
                if (!this.config.silent)
                    this.screen.info("The log will be written to " + this.config.logFilePath);
                this.logStream.on("error", (err) => {
                    this.exit("Error writing to log file: " + err);
                });
            }
            catch (err) {
                this.exit("Error creating log stream: " + err);
            }
        }
    }
    /**
     * Format message for file output (without colors)
     * @private
     */
    _formatMessage(type, message, time) {
        return `[${time || this.toolkit.formatTime()}][${type}] ${this.toolkit.encryptPrivacyContent(this.toolkit.stringify(message))}`;
    }
    /**
     * Write log to file
     * @private
     */
    _log(type, message, time) {
        if (!this.config.logFilePath)
            return;
        if (!this.logStream) {
            if (!this.config.silent)
                this.screen.warning("RLog not initialized, automatic init in progress...");
            this.init();
        }
        this.writeLogToStream(this._formatMessage(type, message, time) + "\n");
    }
    /**
     * Write text to log stream asynchronously
     * @param text - Text to write
     * @returns Promise that resolves when write completes
     */
    writeLogToStream(text) {
        return new Promise((resolve, reject) => {
            if (this.logStream) {
                this.logStream.write(text, (err) => {
                    if (err)
                        reject(err);
                    else
                        resolve();
                });
            }
            else {
                reject(new Error("Log stream not initialized"));
            }
        });
    }
    /**
     * Write an information log to file
     * @param message - Message to log
     * @param time - Optional custom timestamp
     */
    info(message, time) {
        this._log("INFO", message, time);
    }
    /**
     * Write a warning log to file
     * @param message - Message to log
     * @param time - Optional custom timestamp
     */
    warning(message, time) {
        this._log("WARN", message, time);
    }
    /**
     * Write a warning log to file (alias for warning)
     * @param message - Message to log
     * @param time - Optional custom timestamp
     */
    warn(message, time) {
        this._log("WARN", message, time);
    }
    /**
     * Write an error log to file
     * @param message - Message to log
     * @param time - Optional custom timestamp
     */
    error(message, time) {
        this._log("ERROR", message, time);
    }
    /**
     * Write a success log to file
     * @param message - Message to log
     * @param time - Optional custom timestamp
     */
    success(message, time) {
        this._log("SUCCESS", message, time);
    }
    /**
     * Write an exit log to file
     * @param message - Message to log
     * @param time - Optional custom timestamp
     */
    exit(message, time) {
        this._log("EXIT", message, time);
    }
}
// 使用 globalThis 避免类型声明冲突
if (typeof globalThis !== 'undefined') {
    globalThis.__RLOG_EXIT_CONTEXT = null;
}
/**
 * Main Rlog class
 * A comprehensive logging library for Node.js with TypeScript support
 *
 * @example
 * ```typescript
 * import Rlog from 'rlog-js';
 *
 * const rlog = new Rlog({
 *   logFilePath: './logs.txt',
 *   timezone: 'Asia/Shanghai',
 *   enableColorfulOutput: true
 * });
 *
 * rlog.info('Application started');
 * rlog.warn('Warning message');
 * rlog.error('Error occurred');
 * rlog.success('Operation completed');
 * ```
 */
class Rlog {
    /**
     * Create a new Rlog instance
     * @param config - Optional configuration object
     */
    constructor(config) {
        _Rlog_instances.add(this);
        /**
         * Log an information message
         * Outputs to both screen and file
         * @param args - Messages to log
         */
        this.info = __classPrivateFieldGet(this, _Rlog_instances, "m", _Rlog_genApi).call(this, "info");
        /**
         * Log a warning message
         * Outputs to both screen and file
         * @param args - Messages to log
         */
        this.warning = __classPrivateFieldGet(this, _Rlog_instances, "m", _Rlog_genApi).call(this, "warning");
        /**
         * Log a warning message (alias for warning)
         * Outputs to both screen and file
         * @param args - Messages to log
         */
        this.warn = __classPrivateFieldGet(this, _Rlog_instances, "m", _Rlog_genApi).call(this, "warn");
        /**
         * Log an error message
         * Outputs to both screen and file
         * @param args - Messages to log
         */
        this.error = __classPrivateFieldGet(this, _Rlog_instances, "m", _Rlog_genApi).call(this, "error");
        /**
         * Log a success message
         * Outputs to both screen and file
         * @param args - Messages to log
         */
        this.success = __classPrivateFieldGet(this, _Rlog_instances, "m", _Rlog_genApi).call(this, "success");
        this.config = new Config();
        this.config.setConfig(config || {});
        this.toolkit = new Toolkit(this.config);
        this.screen = new Screen(this.toolkit);
        this.toolkit.screen = this.screen;
        this.file = new File(this.toolkit, this.config, this.screen);
        this.exitListeners = [];
        this.keywordPatterns = {
            success: /(success|ok|done|✓)/i,
            warning: /(warn|but|notice|see|problem)/i,
            error: /(error|fail|mistake|problem|fatal)/i,
        };
    }
    /**
     * Display a progress bar in the console
     * @param num - Current progress value
     * @param max - Maximum progress value
     * @example
     * ```typescript
     * for (let i = 0; i <= 100; i++) {
     *   rlog.progress(i, 100);
     *   await sleep(10);
     * }
     * ```
     */
    progress(num, max) {
        const timeheader = `[${this.toolkit.formatTime()}]`;
        const percent = Math.floor(100 * (num / max)) + "%";
        const paddedPercent = " ".repeat(4 - percent.length) + percent;
        const numStr = num.toString();
        const maxStr = max.toString();
        const state = `${" ".repeat(maxStr.length - numStr.length)}${numStr}/${maxStr}`;
        const availableLength = process.stdout.columns -
            timeheader.length -
            6 -
            3 -
            state.length -
            1 -
            paddedPercent.length;
        if (availableLength <= 1) {
            process.stdout.write(`\r${timeheader}[${chalk_1.default.magenta("PROG")}] ${paddedPercent} ${state}`);
        }
        else {
            const doneLength = Math.floor(availableLength * (num / max));
            process.stdout.write(`\r${timeheader}[${chalk_1.default.magenta("PROG")}] [${"|".repeat(doneLength)}${" ".repeat(availableLength - doneLength)}]${paddedPercent} ${state}`);
        }
    }
    /**
     * Exit the program safely
     * Ensures all logs are written before terminating
     * @param message - Exit message
     * @throws {RLogExitError} Special error that triggers exit handler
     * @example
     * ```typescript
     * if (criticalError) {
     *   rlog.exit('Critical error occurred');
     * }
     * ```
     */
    exit(message) {
        const time = this.toolkit.formatTime();
        this.screen.exit(message, time);
        const ExitError = new Error("RLog_EXIT_PROCESS");
        ExitError.isRLogExit = true;
        ExitError.message = message;
        ExitError.time = time;
        globalThis.__RLOG_EXIT_CONTEXT = {
            file: this.file,
            exitListeners: this.exitListeners,
        };
        throw ExitError;
    }
    /**
     * Smart logging function that automatically determines log level
     * Analyzes message content and routes to appropriate log method
     * @param args - Messages to log
     * @example
     * ```typescript
     * rlog.log('Operation successful');  // -> success
     * rlog.log('Warning: disk space low');  // -> warning
     * rlog.log('Error occurred');  // -> error
     * rlog.log('Processing...');  // -> info
     * ```
     */
    log(...args) {
        const message = args.join(this.config.joinChar);
        for (const [key, regex] of Object.entries(this.keywordPatterns)) {
            if (regex.test(message)) {
                this[key](message);
                return;
            }
        }
        this.info(message);
    }
    /**
     * Register a callback to be executed before program exit
     * Useful for cleanup operations, saving state, etc.
     * @param callback - Function to call on exit
     * @example
     * ```typescript
     * rlog.onExit(async () => {
     *   await db.close();
     *   console.log('Database closed');
     * });
     * ```
     */
    onExit(callback) {
        if (typeof callback === "function") {
            this.exitListeners.push(callback);
        }
    }
}
_Rlog_instances = new WeakSet(), _Rlog_genApi = function _Rlog_genApi(key) {
    return (...args) => {
        const message = args.length === 1 ? args[0] : args.join(this.config.joinChar);
        const time = this.toolkit.formatTime();
        this.file[key](message, time);
        this.screen[key](message, time);
    };
};
Rlog.Config = Config;
Rlog.Toolkit = Toolkit;
Rlog.Screen = Screen;
Rlog.File = File;
process.on("uncaughtException", async (err) => {
    const rlogErr = err;
    if (rlogErr.isRLogExit && globalThis.__RLOG_EXIT_CONTEXT) {
        const { file, exitListeners } = globalThis.__RLOG_EXIT_CONTEXT;
        try {
            if (file.logStream) {
                file.exit(rlogErr.message, rlogErr.time);
                if (typeof file.logStream.flush === "function") {
                    file.logStream.flush();
                }
                await new Promise((resolve, reject) => {
                    file.logStream.on("finish", resolve);
                    file.logStream.on("error", reject);
                    file.logStream.end();
                });
            }
            await Promise.all(exitListeners.map((listener) => {
                try {
                    return Promise.resolve(listener());
                }
                catch (e) {
                    return Promise.resolve();
                }
            }));
        }
        catch (e) {
            console.error("Error during log finalization:", e);
        }
        finally {
            globalThis.__RLOG_EXIT_CONTEXT = null;
            process.exit(0);
        }
    }
    else {
        console.error("Uncaught exception:", err);
        process.exit(1);
    }
});
process.on("beforeExit", async () => {
    const ctx = globalThis.__RLOG_EXIT_CONTEXT;
    if (ctx?.file?.logStream) {
        const stream = ctx.file.logStream;
        await new Promise((resolve) => {
            stream.once("finish", resolve);
            stream.end();
        });
    }
});
exports.default = Rlog;
module.exports = Rlog;
module.exports.default = Rlog;
